사장님 회원가입, 식당 등록

bgpark82 opened the pull request:
- [x] 사장님 회원가입
- [x] 식당 등록

peter-alethio committed:
#8 라이브러리 추가

peter-alethio committed:
#8 git ignore 추가 (application-local.yml)

peter-alethio committed:
#8 사장님 회원가입

peter-alethio committed:
#8 식당 등록

bgpark82 commented on src/test/java/com/flab/commerce/domain/owner/OwnerServiceTest.java:
```diff
@@ -0,0 +1,56 @@
+                .passwordConfirm("1234")
+                .name("박병길")
+                .phone("01045808682");
+    }
```
개인적으로 Fixture 관리가 좀 어려운 것 같습니다.
자주 사용하는 Fixture는 위처럼 모든 값들을 다 가지고 재활용해서 사용합니다.
하지만, 중간에 값을 바꿔 사용할 경우가 많은데 그 경우는 어떻게 사용하는지 궁금합니다

개인적으로는 3가지 방법을 사용합니다

1. Builder를 반환하여 사용
```java
    private OwnerRegisterDto.OwnerRegisterDtoBuilder createRequest() {
        return OwnerRegisterDto
                .builder()
                .email("bgpark82@gmail.com")
                .password("1234")
                .passwordConfirm("1234")
                .name("박병길")
                .phone("01045808682");
    }
```
빌더를 반환해서 아래 처럼 값을 변경하여 사용합니다. 
개인적으로 가독성이 좋지 않다고 보지만 이게 그나마 괜찮은 방법이라 생각합니다.
```java
createRequest().email("bgpark@gmail.com").build()
```
2. 필드에 따라 다른 메소드 사용
```java
createRequest()
createRequestWithWrongEmail()
createRequestWithWrongPassword()
```
값에 따라 다른 메소드 명을 사용하는 방법입니다.
이 경우는 메소드 명 보고 파악하기 쉬운데 관리하기가 참 어려운 것 같습니다.

3. 파라미터 변경
```java
createRequest(String email, String name);
createRequest(String email);
```
필요할 때마다 메소드 파라미터로 주입을 해주는 방법입니다. 
기본 Fixture를 만들어 놓고 필요할 때마다 파라미터로 값을 주입하여 사용합니다.
```java
   // before : 기본 Fixture 
   private OwnerRegisterDto createRequest() {
        return OwnerRegisterDto
                .builder()
                .email("bgpark82@gmail.com")
                .password("1234")
                .build();
    }
    createRequest();

   // after : 기본 Fixture 
   private OwnerRegisterDto createRequest(String email, String password) {
        return OwnerRegisterDto
                .builder()
                .email(email)
                .password(password)
                .build();
    }
    createRequest("bgpark82@gmail.com","1234);
```
이 경우 기본 Fixture를 생성하기 위해서 계속 주입을 해줘야 되서 불편한 것 같습니다

```
createRequest(String name);
createRequest(String email);
```
문제는 위와 같은 경우에는 Fixture를 사용할 수 없어서 고민입니다.

3가지 방법 외에 어떤 방법이 있는지 어떤 방법을 주로 사용하는지 궁금합니다.

f-lab-jason commented on src/main/java/com/flab/commerce/domain/owner/OwnerController.java:
```diff
@@ -0,0 +1,22 @@
+    @PostMapping("/user/register")
+    public ResponseEntity register(@Valid @RequestBody OwnerRegisterDto registerRequest) {
+        ownerService.register(registerRequest);
+        return ResponseEntity.noContent().build();
```
가입 성공 response를 `ResponseEntity.ok()...`가 아니라 204로 리턴해주는 이유가 있을까요?

bgpark82 replied:
204을 반환한 이유는 리소스 저장에 성공하고 응답바디를 반환하지 않기 때문에 사용했습니다.
200를 사용하지 않은 이유는 리소스 저장에는 성공했지만 응답 바디가 없기 때문입니다.
201을 사용하지 않은 이유는 Location 헤더에 redirect할 위치가 정해지지 않았기 때문에 사용하지 않았습니다

통상적으로는 200을 사용하나요?

혹시 몰라 찾아봤는데 201 NoContent는 아래 경우 사용하는 것 같습니다
```
리소스를 저장하고 다른 페이지로 이동할 필요없는 경우 (예. 문서 편집)
```
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/204
https://httpwg.org/specs/rfc7231.html#status.204

f-lab-jason replied:
일반적으로 새로 가입시에는 리소스를 생성하고 생성된 리소스를 200으로 반환해주는 케이스가 더 많은 것 같아요. (204가 잘못되었다는 얘기는 아닙니다 😅)

f-lab-jason commented on src/main/java/com/flab/commerce/domain/owner/OwnerService.java:
```diff
@@ -0,0 +1,24 @@
+    private final OwnerMapper ownerMapper;
+
+    public boolean register(OwnerRegisterDto registerRequest) {
+        if (!registerRequest.getPassword().equals(registerRequest.getPasswordConfirm())) {
+            throw new RuntimeException("비밀번호와 확인용 비밀번호가 일치하지 않습니다");
+        }
+
+        if (ownerMapper.findByEmail(registerRequest.getEmail()) != null) {
+            throw new RuntimeException("이미 존재하는 이메일입니다");
+        }
```
기본적으로 User의 입력이 잘못된 경우라고 할 수 있는데 `RuntimeException`이 아닌 `BadInputException`같은걸로 wrapping해서 던져주면 좋을 것 같아요. 400에러를 던져줄 수 있도록 해서요.

bgpark82 replied:
혹시 이런 방식을 말씀하신 걸까요?
https://cheese10yun.github.io/spring-guide-exception/

f-lab-jason replied:
네 남겨주신것도 좋은 예제라고 할 수 있겠네요. 사용자가 받는 exception의 관점에서 RuntimeException은 너무 추상적이고(nested exception의 parent라면 관계없지만) 구체적인형태의 exception으로 돌려주면 더 좋습니다. error message와는 별개로요.

f-lab-jason reviewed (changes requested):
1. 태균님도 마찬가지인데 Rest API 규격을 맞출 계획이면 관련해서 한 번 찾아보시면 좋겠습니다. 예를 들어 Post mapping 자체가 create를 의미하기때문에 `.../register`같은 URL은 추가해주지 않구요, 리소스는 복수형으로 적어줍니다. 따라서 `/users`를 Post로 호출하면 '유저 생성'의 의미를 가진다고 할 수 있어요.
2. 테스트에서 실패 케이스들도 같이 추가해주면 좋겠습니다. 특히 validation 관련 테스트는 하나도 없네요.

peter-alethio committed:
#8 사장님 회원가입 url 변경

peter-alethio committed:
#8 POST 메소드 응답 상태 변경

- 204 -> 200

peter-alethio committed:
#8 비밀번호 확인 로직 제거

프론트엔드에서 비밀번호 확인 처리

peter-alethio committed:
#8 커스텀 예외 추가

- 예외가 계속 추가되면 @ExceptionHandler, @RestControllerAdvisor으로 변경 예정
- 참고: https://spring.io/blog/2013/11/01/exception-handling-in-spring-mvc

peter-alethio committed:
#8 사장님 회원가입 에지 케이스 테스트 추가

peter-alethio committed:
#8 사장님 회원가입 http 요청 추가

f-lab-jason reviewed (approved):
🚢
